arrow-site/docs/docs/patterns/monads/README.md

```diff
@@ -365,318 +348,249 @@ Subsequently, `addStep` is called two times to transfer to `Conference` and then
+fun interface Effect<F> {
+  fun control(): DelimitedScope<F>
 }
 ```
 
-See that, instead of `WorkflowThatReturns<A>`, our containers are called `Kind<F, A>`, where `F` is the generic parameter of the container and `A` the generic parameter of the content. We talk more about them in the [glossary]({{ '/patterns/glossary/#type-constructors' | relative_url }}).
-
-The constructor `just` is used to put an object into a container `Kind<F, A>`, and `flatMap` is used to replace one contained object with another contained object.
+We can then define specific effects for our data types that implement Monad bind.
 
-We can think of `flatMap` as a combination of `map` and `flatten`, as defined by the following signature:
+```kotlin:ank
+import arrow.continuations.Effect
 
-```kotlin
-fun <F, A, B> Kind<F, A>.map(f: (A) ->  B): Kind<F, B> // Inherited from Functor
-
-fun <F, A> Kind<F, Kind<F, A>>.flatten(): Kind<F, A>
-
-container
-  .map { x -> just(x + 1) } // Kind<F, Kind<F, Int>>
-  .flatten() // Kind<F, Int>
+fun interface NullableEffect<A> : Effect<A?> {
+  suspend fun <B> B?.bind(): B =
+    this ?: control().shift(null)
+}
 ```
 
-Even though I spent quite some time with examples, I expect you to be slightly confused at this point. That's ok.
+Monad `bind` is then implemented by shifting out of the context with `null` if the value being bound happens to be null or yielding the value by returning it if not null.
 
-Keep going and let's have a look at several sample implementations of Monad pattern.
+`control().shift(value)` can be seen as a functional throw. Once we pass a value the context of our Effect will exit with `value`.
 
-### Option
-
-My first example was with nullable `?`. The full pattern containing either 0 or 1 instance of some type is called Option (it might have a value, but might not).
-
-Option is another approach to dealing with "no value" value, an alternative to the concept of null. You can read more about [`Option`]({{ '/apidocs/arrow-core-data/arrow.core/-option/' | relative_url }}) to see how Arrow implements it.
-
-When null is not allowed, any API contract gets more explicit: Either you return type `T` and it's always going to be filled, or you return `Option<T>`.
-The client will see that Option type is used, so it will be forced to handle the case of absent value.
+With `control().shift` based on the `Effect` interface we can implement monad bind short-circuiting and other patterns supported by the Continuation monad for arbitrary data types.
 
-Given an imaginary repository contract (which does something with customers and orders):
-
-```kotlin
-interface OptionAwareRepository {
-    fun getCustomer(id: Int): Option<Customer>
-
-    fun getAddress(id: Int): Option<Address>
-
-    fun getOrder(id: Int): Option<Order>
+```kotlin:ank
+object nullable {
+  suspend operator fun <A> invoke(func: suspend NullableEffect<*>.() -> A?): A? =
+    Effect.restricted(eff = { NullableEffect { it } }, f = func, just = { it })
```

Comment:
I think this is incorrect. It's using `restricted` but it exposing an `suspend` function.